contracts/cw-auth-manager/src/contract.rs

```diff
@@ -0,0 +1,110 @@
+use cosmwasm_std::{Addr, Binary, CosmosMsg, Deps, DepsMut, Empty, Env, MessageInfo, Reply, Response, StdResult, to_binary};
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cw2::set_contract_version;
+
+use crate::{
+    error::ContractError,
+    msg::{ExecuteMsg, InstantiateMsg, QueryMsg},
+    state::{AUTHORIZATIONS, CONFIG, Config}
+};
+use crate::msg::IsAuthorizedResponse;
+use crate::state::Authorization;
+
+const CONTRACT_NAME: &str = "crates.io:cw-auth-manager";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    _msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    let config = Config { dao: info.sender.clone() };
+    let empty: Vec<Authorization> = vec![];
+    CONFIG.save(deps.storage, &config)?;
+    AUTHORIZATIONS.save(deps.storage, &info.sender, &empty)?;
+    Ok(Response::default().add_attribute("action", "instantiate"))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::AddAuthorization { auth_contract } => execute_add_authorization(deps, env, info, auth_contract),
+    }
+}
+
+pub fn execute_add_authorization(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    address: String,
```

Comment:
have you checked out the cw-named-groups contract in this repo? Was originally written with the intention of prescribing sets of permissions to different groups of addresses. I'm wondering if we could use that in combination with this contract to assign a role or authorization to a group, and if a user gets added to a group they automatically receive the authorizations of that group

Reply:
I started experimenting with integrating the authorizations into groups by storing an auth manager and adding an `authorize()` method directly in https://github.com/nicolaslara/dao-contracts/blob/5ccda74e48cc4ee36861100bae9a7ba2240fe7ce/contracts/cw-named-groups/src/state.rs#L13. I don't think this will work, as the authorizations would need to behave differently for groups and addresses. 

This would work if we expect authorizations to be able to handle both queries (`Authorize` and `AuthorizeGroup`), but I think this would lead to a lot of repeated code and I'm concerned about implementing auths that check for groups based only on their names. Seems like it may open to security issues in the future. 

I think an alternative may be to move the group checks to the `cw-auth-manager` contract, but I'm still not sure what the cleanest way to do this would be. I'll keep thinking about this and push a suggestion for review when I have something concrete.

Reply:
@janitachalam Just pushed an implementation of something like what I described above. I'm not particularly happy with this, since there is a lot of manual work that needs to be done on the underlaying auth contracts (whitelist, being the example auth contract in this case).

The ideal way to do this would be to make auths generic by referencing groups and addresses by the same unique identifier. I think the easiest way to do this is to assign an `Addr` to each group, but that would require modfying `cw-named-groups` considerably and potentially introduce a second level of indirection (where each group has its own contract). Besides the complexity and need for a migration of existing groups, I'm not sure what performance impact this would have.

Reply:
changed my previous implementation here so that auths are stored as `String` instead of `Addr`. I think this gives us the flexibility that I'd want. The contracts will behave the same if they're authorizing an auth or a group (it's just a string comparison!). 

My only concern with this is that it may open to unexpected semantics. For example: If someone is allowed to create a group whose name is the same as the Addr of someone with higher permissions we would end up with a permission escalation issue.